### PR TITLE
Adding tutorial about cache and also ability to pass a cache directly to the tournament

### LINKED ISSUE #547: Write tutorial for cache
To go in the "Further capabilities in the library" section.

### PR DESCRIPTION
**Pull request by drvinceknight**
Closes #547 

**This is on top of #546, so that should be merged/reviewed first. (Merging this will merge #546)**

Note that to make things in line with the match and the Moran process I've added the ability to pass a cache directly to the `Tournament` class. So the following works (as shown in the tutorial):

```
>>> cache = axl.DeterministicCache(filename)  # Read in some prebuilt cache
>>> players = [axl.GoByMajority(), axl.Alternator(), axl.Cooperator()]
>>> tournament = axl.Tournament(players, turns=200, repetitions=5, deterministic_cache=cache)
>>> results = tournament.play()
```

I haven't gotten rid of the `prebuilt_cache` argument as I didn't want to get in to the `TournamentManager` (I think that's where it's used). That could potentially be made simpler with this though?

### CODE REVIEW

File: docs/tutorials/advanced/using_the_cache.rst [hunk 1]
```diff
@@ -0,0 +1,99 @@
+.. _using-the-cache:
+
+Using the cache
+===============
+
+Whilst for stochastic strategies every repetition of a Match will give a
+different result. For deterministic strategies, when there is no noise there is
```
Comment (meatballs) [inline docs/tutorials/advanced/using_the_cache.rst]: Whilst for stochastic strategies, every repetition of a Match will give a different result, for deterministic strategies, .....

File: docs/tutorials/advanced/using_the_cache.rst [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    >>> time_with_empty_cache = timeit.timeit(run_match, number=500)
+
+We can take a look at the cache::
+
+    >>> cache  # doctest: +ELLIPSIS
+    {(<class 'axelrod.strategies.gobymajority.GoByMajority'>, <class 'axelrod.strategies.alternator.Alternator'>, 200): [('C', 'C'), ..., ('C', 'D')]}
+    >>> len(cache)
+    1
+
+Note, that is maps triplets of players and match lengths to interactions.
```
Comment (meatballs) [inline docs/tutorials/advanced/using_the_cache.rst]: typo here, I think

Comment (meatballs) [inline docs/tutorials/advanced/using_the_cache.rst]: Might be better as:

Note: This maps a triplet of 2 player classes and the match length to the resulting interactions.

**Comment by meatballs** [comment]
Just minor wording comments on the docs from me. Everything else looks good.

**Comment by drvinceknight** [comment]
> Just minor wording comments on the docs from me. Everything else looks good.

Think I've picked them all up. Thanks :) 👍

**Comment by drvinceknight** [comment]
Just walking the puppy and figured out something that's not quite right with this. Will be a two second fix.

**Comment by drvinceknight** [comment]
I wasn't timing the run without the cache properly (`timeit` was in fact just running that once without a cache and then every other time with a cache). Have fixed this now (and it makes for an even more impressive time difference :)).

**Pull request merged**